go-controller/pkg/ovn/ipallocator/allocator/bitmap.go

```diff
@@ -234,3 +247,27 @@ func (contiguousScanStrategy) AllocateBit(allocated *big.Int, max, count int) (i
 }
 
 var _ bitAllocator = contiguousScanStrategy{}
+
+// roundRobinScanStrategy tries to allocate starting at 0 and scans forward looking for
+// the next available address, wrapping the range if necessary. It caches the
+// last allocation and will try to allocate all other addresses before re-allocating
+// the most recently allocated address.
+type roundRobinScanStrategy struct {
+	lastAllocated int
+}
+
+func (wss *roundRobinScanStrategy) AllocateBit(allocated *big.Int, max, count int) (int, bool) {
+	if count >= max {
+		return 0, false
+	}
+	for i := 0; i < max; i++ {
+		at := (wss.lastAllocated + i) % max
+		if allocated.Bit(at) == 0 {
+			wss.lastAllocated = at
```

Comment:
@dcbw do we have an issue of off-by-1 here? Should we start from` lastAllocated + 1` bit for new allocation?

```
func main() {
	_, cidr, _ := net.ParseCIDR("192.168.1.0/29")
	subnetRange, err := ipam.NewAllocatorCIDRRange(cidr, func(max int, rangeSpec string) (allocator.Interface, error) {
		return allocator.NewRoundRobinAllocationMap(max, rangeSpec), nil
	})
	if err != nil {
		fmt.Printf("failed creating allocator\n")
		return
	}
	for i := 0; i < 8; i++ {
		ip, err := subnetRange.AllocateNext()
		if err != nil {
			fmt.Printf("failed to allocated an ip - %d\n", i)
			continue
		}
		fmt.Printf("allocated IP: %s\n", ip.String())

		err = subnetRange.Release(ip)
		if err != nil {
			fmt.Printf("failed to release the IP: %s\n", ip.String())
			continue
		}
		// fmt.Printf("released IP: %s\n", ip.String())
	}
}
```
running the above program results in 

allocated IP: 192.168.1.1
allocated IP: 192.168.1.1
allocated IP: 192.168.1.1
allocated IP: 192.168.1.1
allocated IP: 192.168.1.1
allocated IP: 192.168.1.1
allocated IP: 192.168.1.1
allocated IP: 192.168.1.1

I was hoping it would be .1, .2., .3, .4, .5,.6... and again  .1, .2, .3, .4,...

Reply:
good catch, the for loop has to start in 1 instead of 0, otherwise it tries to allocate first in the last allocated position, that in this case was released and is free

Reply:
@girishmg @aojea fixed, with a testcase added to catch that.